sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java

```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
-        twins
-            .forEach((twinId, twinContent) -> {
-                // Call APIs to delete all relationships.
-                client.listRelationships(twinId, BasicRelationship.class)
-                    .doOnComplete(deleteRelationshipsSemaphore::release)
-                    .doOnError(throwable -> {
-                        if (throwable instanceof ErrorResponseException && ((ErrorResponseException) throwable).getResponse().getStatusCode() == HttpStatus.SC_NOT_FOUND) {
-                            deleteRelationshipsSemaphore.release();
-                        } else {
-                            System.err.println("List relationships error: " + throwable);
-                        }
-                    })
-                    .subscribe(
-                        relationship -> client.deleteRelationship(twinId, relationship.getId())
-                            .subscribe(
-                                aVoid -> System.out.println("Found and deleted relationship: " + relationship.getId()),
-                                throwable -> System.err.println("Delete relationship error: " + throwable)
-                            ));
-
-                // Call APIs to delete any incoming relationships.
-                client.listIncomingRelationships(twinId)
-                    .doOnComplete(deleteRelationshipsSemaphore::release)
-                    .doOnError(throwable -> {
-                        if (throwable instanceof ErrorResponseException && ((ErrorResponseException) throwable).getResponse().getStatusCode() == HttpStatus.SC_NOT_FOUND) {
-                            deleteRelationshipsSemaphore.release();
-                        } else {
-                            System.err.println("List incoming relationships error: " + throwable);
-                        }
-                    })
-                    .subscribe(
-                        incomingRelationship -> client.deleteRelationship(incomingRelationship.getSourceId(), incomingRelationship.getRelationshipId())
-                            .subscribe(
-                                aVoid -> System.out.println("Found and deleted incoming relationship: " + incomingRelationship.getRelationshipId()),
-                                throwable -> System.err.println("Delete incoming relationship error: " + throwable)
-                            ));
+        // Once the async API terminates (either successfully, or with an error), the latch count is decremented, or the semaphore is released.
+        for (Map.Entry<String, String> twin : twins.entrySet()) {
+            String twinId = twin.getKey();
+
+            List<BasicRelationship> relationshipList = new ArrayList<>();
+            Semaphore listRelationshipSemaphore = new Semaphore(0);
+            Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
+            CountDownLatch deleteTwinsLatch = new CountDownLatch(1);
```

Comment:
Unfortunate that you have to know ahead of time how many items you are counting down. This presents a "magic" number, which can be confusing to someone else reviewing the code who doesn't realize the significance (this will be used to asynchronously delete 1 dt).

Also, in C# we have a general rule of not declaring a variable until it is needed (so it is more contextual). Does Java have a different standard of all variables at the top?

FWIW, I think this could at least use a comment.

Reply:
We need to initialize these latches and semaphore within the scope that they are referenced in; but yes, I agree, the count can be confusing to understand.
I'll add some more comments around this.